### PR TITLE
[feat]: [CDS-47024] ASG NG Infrastructure Implementation

### PR DESCRIPTION
**Pull request by lovish1999**
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/41241)

### CODE REVIEW

**Review by vitsafronovici — APPROVED** [review]
> Reviewed 12 of 19 files at r1, 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41241)* status: 19 file(s) to review. 1 of 2 approvals obtained.

File: 125-cd-nextgen/src/main/java/io/harness/cdng/infra/InfrastructureValidator.java [hunk 1]
```diff
@@ -256,6 +261,16 @@ private void validateTanzuApplicationServiceInfrastructure(TanzuApplicationServi
     }
   }
 
+  private void validateAsgInfrastructure(AsgInfrastructure infrastructure) {
+    if (!hasValueOrExpression(infrastructure.getConnectorRef())) {
+      throw new InvalidArgumentsException(Pair.of("connectorRef", "cannot be empty"));
+    }
+
+    if (null == infrastructure.getRegion()) {
```
Comment (yogesh-chauhan) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/infra/InfrastructureValidator.java:269]: should we check for empty as well ? also what about hasValueOrExpression method ?

File: 125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/AbstractInfrastructureTaskExecutableStep.java [hunk 2]
```diff
@@ -649,6 +650,13 @@ void validateConnector(Infrastructure infrastructure, Ambiance ambiance, NGLogCa
           ConnectorType.SPOT.name()));
     }
 
+    if (InfrastructureKind.ASG.equals(infrastructure.getKind())
+        && !(connectorInfo.get(0).getConnectorConfig() instanceof AwsConnectorDTO)) {
+      throw new InvalidRequestException(format("Invalid connector type [%s] for identifier: [%s], expected [%s]",
+          connectorInfo.get(0).getConnectorType().name(), infrastructure.getConnectorReference().getValue(),
+          ConnectorType.AWS.name()));
+    }
+
```
Comment (yogesh-chauhan) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/AbstractInfrastructureTaskExecutableStep.java:659]: Make sure that `InfrastructureTaskExecutableStepV2` code flow has this logic as well

**Review by yogesh-chauhan — APPROVED** [review]

File: 125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/AbstractInfrastructureTaskExecutableStep.java (hunk 2 above)
Comment (lovish1999) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/AbstractInfrastructureTaskExecutableStep.java:659]: Okay.

**Comment by lovish1999** [comment]
**Description:**
I have added the Infrastructure implementation for ASG NG Swimlane. It is based on the YAML given in https://harness.atlassian.net/wiki/spaces/CDNG/pages/21231173771/AWS+ASG+YAML+Proposal+1

File: 125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/AbstractInfrastructureTaskExecutableStep.java (hunk 2 above)
Comment (lovish1999) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/AbstractInfrastructureTaskExecutableStep.java:659]: The _executeSync_ method in _InfrastructureTaskExecutableStepV2_ calls this - _infrastructureValidator.validate(spec)_   - this then executes the required validation logic

**Review by yogesh-chauhan — APPROVED** [review]

**Review by lovish1999 — COMMENTED** [review]
> Reviewed 9 of 19 files at r1, 6 of 7 files at r2, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41241)* status: 19 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained.

**Review by lovish1999 — COMMENTED** [review]
LGTM

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41241#-:-NIpkdhw-ArvaTS5IMbU:bb6ywmx)* status: 19 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained.

**Review by vitsafronovici — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41241)* status: 19 file(s) to review. 2. unresolved discussions.

**Review by lovish1999 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41241)* status: 19 file(s) to review. 2. unresolved discussions. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/AbstractInfrastructureTaskExecutableStep.java` line 659 at r2](https://reviewable.io/reviews/harness/harness-core/41241#-NIgLMrRBe1hCGGq4wRq-r2-659:-NIq3kCZ7kiPR0AaaILE:b3kqx) ([raw file](https://github.com/harness/harness-core/blob/a083734d76f06e49e565539c5f5603391bde4c3c/125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/AbstractInfrastructureTaskExecutableStep.java#L659)):*
<details><summary><i>Previously, lovish1999 (Lovish Bansal) wrote…</i></summary><blockquote>

The _executeSync_ method in _InfrastructureTaskExecutableStepV2_ calls this - _infrastructureValidator.validate(spec)_   - this then executes the required validation logic
</blockquote></details>

.

**Pull request merged**